#+TITLE: CONTACT
#+AUTHOR: Pjotr Prins

* CONTACT and SUPPORT

COVID-19 PubSeq was created by a group of [[https://github.com/arvados/bh20-seq-resource/graphs/contributors][bioinformaticians]] and
software developers working at leading institutes (see sponsors below)
with the goal of making online analysis available to everyone. You can
talk with us directly on a [[https://gitter.im/arvados/pubseq?utm_source=share-link&utm_medium=link&utm_campaign=share-link][gitter channel]]. We are open to improving
work flows and analysis.

** Oxford Nanopore Analysis

@@html: <img class="img-right" src="static/image/oxford-nanopore.jpg" />@@

We run [[https://en.wikipedia.org/wiki/Oxford_Nanopore_Technologies][Oxford Nanopore]] ourselves. It is an exaciting technology
because it gives us an accurate SARS-CoV-2 sequencer for a few
thousand dollars which can be run in a living room! With PubSeq we aim
to make it easy to analyse Nanopore material using our *free* Cloud
infrastructure. If you need help in using the online workflows don't
hesitate to contact us.

@@html: </br></br></br>@@

** Data from other sequencers

We accept FASTA sequences of SARS-CoV-2. Simply upload them using the
web form and/or REST API. No sign-up required! If you have large scale
short read data and require support we can discuss that. We also run
Illumina sequencing in several places.

** Professional support

To use COVID-19 PubSeq solutions for professional purposes you can
contact Boston based [[mailto:info@curii.com][Curii]], the creators of [[https://arvados.org/][Arvados]], directly.

COVID-19 is built on Arvados using CWL workflows.

** E-mail

For questions feel free to write directly to [[mailto:pjotr.public821@thebird.nl][Pjotr Prins]].
